### ISSUE TITLE
ForwardAccumulator: The dtype of the watched primal must be floating

### ISSUE THREAD

**Issue by nkoep** [tensorflow/tensorflow#56415]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Others

### Source

binary

### Tensorflow Version

2.9.1

### Custom Code

No

### OS Platform and Distribution

Linux

### Python version

3.9

### Current Behaviour?

```shell
Hi! I noticed that `tf.ForwardAccumulator` is currently issuing a warning when it is used to compute a jvp of a real-valued function for complex input. I'm not sure if this is a historic artifact since the result seems correct despite the warning as the code example below demonstrates. I would appreciate if someone could clarify why the warning exists.

The example below implements two Hessian-vector product (hvp) functions, one based on reverse-over-reverse mode using two `tf.GradientTape` instances (i.e., the gradient of the directional derivative), and one based on forward-over-reverse mode using `tf.ForwardAccumulator` and `tf.GradientTape`. The latter issues the warning `WARNING:tensorflow:The dtype of the watched primal must be floating (e.g. tf.float32), got tf.complex64` which originates here: https://github.com/tensorflow/tensorflow/blame/d8ce9f9c301d021a69953134185ab728c1c248d3/tensorflow/python/eager/forwardprop.py#L399-L402.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf


def hessian_vector_product_forward_over_reverse(function, primals, tangents):
    with tf.GradientTape() as tape, tf.autodiff.ForwardAccumulator(
        primals, tangents
    ) as accumulator:
        for primal in primals:
            tape.watch(primal)
        gradients = tape.gradient(function(*primals), primals)
    return accumulator.jvp(gradients)


def hessian_vector_product_reverse_over_reverse(function, primals, tangents):
    with tf.GradientTape() as hvp_tape, tf.GradientTape() as gradient_tape:
        for primal in primals:
            gradient_tape.watch(primal)
            hvp_tape.watch(primal)
        gradients = gradient_tape.gradient(function(*primals), primals)
        dot_product = sum(
            [
                tf.math.real(
                    tf.tensordot(
                        tf.math.conj(gradient), tangent, axes=gradient.ndim
                    )
                )
                for gradient, tangent in zip(gradients, tangents)
            ]
        )
    return hvp_tape.gradient(dot_product, primals)


if __name__ == "__main__":
    tf.random.set_seed(42)

    def function(x):
        return tf.math.real(tf.reduce_sum(x**2))

    n = 10
    x = tf.complex(tf.random.normal((n,)), tf.random.normal((n,)))
    v = tf.complex(tf.random.normal((n,)), tf.random.normal((n,)))

    print("Expected hvp:", (2 * tf.math.conj(v),))
    print(
        "Forward-over-reverse hvp:",
        hessian_vector_product_forward_over_reverse(function, (x,), (v,)),
    )
    print(
        "Reverse-over-reverse hvp:",
        hessian_vector_product_reverse_over_reverse(function, (x,), (v,)),
    )
```

### Relevant log output

```shell
...
Expected hvp: (<tf.Tensor: shape=(10,), dtype=complex64, numpy=
array([-1.1181947 -1.3129735j, -1.0694427 +0.8261034j,
        4.7460666 -0.6799501j, -3.1451862 +2.0112543j,
        1.6110111 -1.4053247j, -1.6677539 +2.8017285j,
        0.61222446+1.7956151j,  4.532099  +0.6971235j,
        0.5712828 +1.9178125j, -3.1072311 -2.3896964j], dtype=complex64)>,)
WARNING:tensorflow:The dtype of the watched primal must be floating (e.g. tf.float32), got tf.complex64
Forward-over-reverse hvp: (<tf.Tensor: shape=(10,), dtype=complex64, numpy=
array([-1.1181947 -1.3129735j, -1.0694427 +0.8261034j,
        4.7460666 -0.6799501j, -3.1451862 +2.0112543j,
        1.6110111 -1.4053247j, -1.6677539 +2.8017285j,
        0.61222446+1.7956151j,  4.532099  +0.6971235j,
        0.5712828 +1.9178125j, -3.1072311 -2.3896964j], dtype=complex64)>,)
Reverse-over-reverse hvp: (<tf.Tensor: shape=(10,), dtype=complex64, numpy=
array([-1.1181947 -1.3129735j, -1.0694427 +0.8261034j,
        4.7460666 -0.6799501j, -3.1451862 +2.0112543j,
        1.6110111 -1.4053247j, -1.6677539 +2.8017285j,
        0.61222446+1.7956151j,  4.532099  +0.6971235j,
        0.5712828 +1.9178125j, -3.1072311 -2.3896964j], dtype=complex64)>,)
```
</details>

**Comment by sushreebarsa** [comment]
@chunduriv I was able to replicate the issue on colab, please find the [gist](https://colab.research.google.com/gist/sushreebarsa/be9b6de6fe107a624828a3149606b255/56415.ipynb) here. Thank you!

**Comment by sachinprasadhs** [comment]
Yes, for Gradients calculation as per the warning it expects watched tensor to be of type float, and Tensorflow does not automatically do casting. 
Check the example [here](https://www.tensorflow.org/guide/autodiff#3_took_gradients_through_an_integer_or_string) for details. Thanks!

**Comment by nkoep** [comment]
The link talks about dtypes for which the gradient has no sensible
definition. TF supports complex autodiff in general though. In my example
the warning is only raised when the I use tf.ForwardAccumulator, using
tf.GradientTape it does not. In both cases, the gradient (or hvp I should
say) is correct though in the sense of the conjugate Wirtinger derivative
that TF uses for real-valued functions of complex arguments; the gradient
is not None as in your example.

I'm just wondering why forward mode autodiff issues a dtype warning while
reverse mode does not. I suspect this is a historical artifact.
GradientTape.watch calls backprop_util.IsTrainable e.g. to determine
whether the dtype is supported (see
https://github.com/tensorflow/tensorflow/blob/d8ce9f9c301d021a69953134185ab728c1c248d3/tensorflow/python/eager/backprop.py#L909
and
https://github.com/tensorflow/tensorflow/blob/d8ce9f9c301d021a69953134185ab728c1c248d3/tensorflow/python/eager/backprop_util.py#L56)
which explicitly allows complex dtypes. ForwardAccumulator._watch only
checks whether the dtype is floating (
https://github.com/tensorflow/tensorflow/blob/d8ce9f9c301d021a69953134185ab728c1c248d3/tensorflow/python/eager/forwardprop.py#L399
).

**Comment by aniruthraj** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
